lib/utils/patterns.ts

```diff
@@ -0,0 +1,228 @@
+		return form.ask({
+			message: 'Device Type',
+			type: 'list',
+			choices: deviceTypes
+		})
+	});
+}
+
+export function confirm(yesOption: string, message: string, yesMessage: string) {
+	return Promise.try(function () {
+		if (yesOption) {
+			if (yesMessage) { console.log(yesMessage); }
+			return true;
+		}
+
+		return form.ask({
+			message,
+			type: 'confirm',
+			default: false
+		});
+	}).then(function(confirmed) {
+		if (!confirmed) {
+			throw new Error('Aborted');
+		}
+	})
+}
+
+export function selectApplication(filter: (app: ResinSdk.Application) => boolean) {
+	resin.models.application.hasAny().then(function(hasAnyApplications) {
+		if (!hasAnyApplications) {
+			throw new Error('You don\'t have any applications');
+		}
+
+		return resin.models.application.getAll();
+	})
+	.filter(filter || _.constant(true))
+	.then(applications => {
+		return form.ask({
+			message: 'Select an application',
+			type: 'list',
+			choices: _.map(applications, application =>
+				({
+					name: `${application.app_name} (${application.device_type})`,
+					value: application.app_name
+				})
+		)});
+	})
+}
+
+export function selectOrCreateApplication() {
+	return resin.models.application.hasAny().then((hasAnyApplications) => {
+		if (!hasAnyApplications) return;
+
+		return resin.models.application.getAll().then((applications) => {
+			let appOptions: { name: string, value: string | null }[];
+			appOptions = _.map(applications, application => ({
+				name: `${application.app_name} (${application.device_type})`,
+				value: application.app_name
+			}));
+
+			appOptions.unshift({
+				name: 'Create a new application',
+				value: null
+			});
+
+			return form.ask({
+				message: 'Select an application',
+				type: 'list',
+				choices: appOptions
+			});
+		});
+	}).then((application) => {
+		if (application != null) return application;
+
+		return form.ask({
+			message: 'Choose a Name for your new application',
+			type: 'input',
+			validate: validation.validateApplicationName
+		});
+	})
+}
+
+export function awaitDevice(uuid: string) {
+	return resin.models.device.getName(uuid)
+	.then((deviceName) => {
+		const spinner = new visuals.Spinner(`Waiting for ${deviceName} to come online`);
+
+		const poll = (): Promise<void> => {
+			return resin.models.device.isOnline(uuid)
+			.then(function(isOnline) {
+				if (isOnline) {
+					spinner.stop();
+					console.info(`The device **${deviceName}** is online!`);
+					return;
+				} else {
+					// Spinner implementation is smart enough to
+					// not start again if it was already started
+					spinner.start();
+
+					return Promise.delay(3000).then(poll);
+				}
+			});
+		}
+
+		console.info(`Waiting for ${deviceName} to connect to resin...`);
+		return poll().return(uuid);
+	});
+}
+
+export function inferOrSelectDevice(preferredUuid: string) {
+	return resin.models.device.getAll()
+	.filter((device: ResinSdk.Device) => device.is_online)
+	.then((onlineDevices: ResinSdk.Device[]) => {
```

Comment:
NIT: This could be changed to
```ts
.filter<ResinSdk.Device>(device => device.is_online)
.then(onlineDevices) => {
```
and the `.then` would properly infer the type.

Reply:
Fixed